### ISSUE TITLE
Avoid using --insecure during bootstrap

### ISSUE THREAD

**Issue by shellwhale** [siderolabs/talos#9521]
Hello.

From my understanding the initial configuration, the bootstrap phase, is not authenticated. 
This means this is a Trust On First Use [(TOFU)](https://en.wikipedia.org/wiki/Trust_on_first_use) authentification scheme, which is vulnerable to Man-in-the-middle attacks.

Is there a way to embed a certificate inside of the image, before the initial configuration that happens over the network?
That way we could get rid of the `--insecure` flag.

**Comment by frezbo** [comment]
bootstrap API requires proper certificate, initial `apply-config` can only use `--insecure`

**Comment by shellwhale** [comment]
> bootstrap API requires proper certificate, initial `apply-config` can only use `--insecure`

Yes that's why I'm asking, it feels great to use a insecure authentification scheme right from the start. 

Again, can't I generate myself a certificate, embed it in the image somehow, maybe as a step from factory.talos.dev or using the imager?

**Comment by shellwhale** [comment]
Anyone with network access can configure the machine, or impersonate it, isn't that an issue?

**Comment by shellwhale** [comment]
There, 

https://www.talos.dev/v1.8/learn-more/image-factory/#schematics
https://www.talos.dev/v1.8/talos-guides/install/boot-assets/#image-factory

I feel this is where there should be an explanation on how to setup a custom certificate, at the image level, not at the network  level. But I can't find any.

**Comment by smira** [comment]
> Is there a way to embed a certificate inside of the image, before the initial configuration that happens over the network?
> That way we could get rid of the `--insecure` flag.

If you follow console output of Talos, you would notice that it prints the fingerprint of its own certificate which you can use with `--insecure` apply config. This is already implemented.

You can submit machine configuration to Talos via many other methods as well, which have their own pros and cons, but if you worry about man-in-the-middle specifically, use the fingerprint shown in the console.

**Comment by shellwhale** [comment]
@smira but isn't the fingerprint the same for everyone?

**Comment by shellwhale** [comment]
> > Is there a way to embed a certificate inside of the image, before the initial configuration that happens over the network?
> > That way we could get rid of the `--insecure` flag.
> 
> If you follow console output of Talos, you would notice that it prints the fingerprint of its own certificate which you can use with `--insecure` apply config. This is already implemented.

So this means that the CLI client trusts server certificates signed by some certificate authority, the private key associated with that CA must be stored on the Talos image itself if the signing does happen at the server boot time. 

Is that what happens?

**Comment by TomyLobo** [comment]
The fingerprint on the console solves the MITM problem.
But as far as I can tell, while Talos is in maintenance mode, anyone can it to install anything, including malware, without authentication, so MITM isn't even necessary :)

**Comment by rothgar** [comment]
If you are worried about someone on your network sending a malicious machine config to a node in maintenance mode then you should use one of the alternative methods for configuring the machine or use Omni with siderolink which creates a secure wireguard connection and doesn't accept insecure connections.
